### PR TITLE
logging enhancements

### PR DESCRIPTION
**Pull request by ClaudiaJ**
Changes formatter to conform to write [ErrorEvent](https://cloud.google.com/error-reporting/reference/rest/v1beta1/ErrorEvent) and [LogEntry](https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry) compatible payloads to be (hopefully) parsed by Stackdriver and Error Reporting dashboard.

Adds and enhances middleware.

Logging middleware for HTTP and gRPC are intended to come first in chain, and provides a logrus entry into context (see [ctxlogrus](https://pkg.go.dev/github.com/grpc-ecosystem/go-grpc-middleware@v1.2.2/logging/logrus/ctxlogrus)). This entry can be extracted and used directly, or you can append fields to be logged downstream and after the middlewares handle the request.

Recovery middleware recovers from panic within the lifecycle of an HTTP or gRPC request, and responds with an internal server error, gRPC Status body (marshaled to JSON in the case of HTTP response), and status details having a unique RequestID that can be then correlated back to our logs and related to trace identifiers throughout our logs. If the recovery middlewares are accidentally chained before the logging middlewares, it just won't log, but will still respond to the user without crashing the server.

Adds a janky formatter in case errors logged `WithError` were produced upstream with [`github.com/pkg/errors.WithStack`](https://pkg.go.dev/github.com/pkg/errors#WithStack), providing the stacktrace into a field (according to the docs) supported by Stackdriver, but use of `stackTrace` field takes precedence (e.g. as used in panic recovery middleware).

Adds log filtering options and default filters on healthcheck and gRPC reflection to cut down on non-transactional cruft.

Parses out and logs trace id, span id, and trace sampled fields into resulting log entry as defined by GCP's docs, should enable us to correlate entire request lifecycle to logs generated by a request so long as all requests are updated to use the context logger.

gRPC error responses (with details, if any) are logged, internal server errors are logged as error. Deadline, peer address, user agent, are all read from the context on incoming requests and logged, and without also writing an implementation of [`stats.RPCStats`](https://pkg.go.dev/google.golang.org/grpc@v1.36.0/stats), I'm just logging the request duration with a start/end time.

TODO: global logger ID vs request-scoped logger ID
TODO: always move special fields to parent context from data for consistency (e.g. `context.data.httpRequest` s `context.httpRequest`
TODO: more correct cloud trace integration:
- https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver
Should be able to pull the current trace span from out of context(?) instead of operating on header directly, and rely on propagator for header parsing.

### CODE REVIEW

**Comment by ClaudiaJ** [comment]
`go test -v` to see this middleware in action. Examples*:

```json
=== RUN   TestServerSuite/TestGood
{
        "logName": "projects/test-project/logs/logging-test",
        "timestamp": "2021-02-28T02:34:08.664268Z",
        "message": "some ping",
        "severity": "INFO",
        "context": {
                "data": {
                        "custom_field": "custom_value",
                        "grpcRequest": {
                                "method": "/mwitkow.testproto.TestService/Ping",
                                "userAgent": "grpc-go/1.36.0",
                                "peer": "tcp://127.0.0.1:61373",
                                "deadline": "2021-02-27T21:34:11.664148-05:00"
                        }
                }
        },
        "sourceLocation": {
                "file": "google.golang.org/grpc@v1.36.0/server.go",
                "line": 1217,
                "function": "(*Server).processUnaryRPC"
        },
        "logging.googleapis.com/trace": "projects/test-project/traces/105445aa7843bc8bf206b12000100000",
        "logging.googleapis.com/spanId": "1",
        "logging.googleapis.com/traceSampled": true
}
{
        "logName": "projects/test-project/logs/logging-test",
        "timestamp": "2021-02-28T02:34:08.664388Z",
        "message": "served RPC /mwitkow.testproto.TestService/Ping",
        "severity": "INFO",
        "context": {
                "data": {
                        "custom_field": "custom_value",
                        "grpcRequest": {
                                "method": "/mwitkow.testproto.TestService/Ping",
                                "userAgent": "grpc-go/1.36.0",
                                "peer": "tcp://127.0.0.1:61373",
                                "deadline": "2021-02-27T21:34:11.664148-05:00",
                                "duration": "0.000142023s"
                        }
                }
        },
        "sourceLocation": {
                "file": "google.golang.org/grpc@v1.36.0/server.go",
                "line": 1217,
                "function": "(*Server).processUnaryRPC"
        },
        "logging.googleapis.com/trace": "projects/test-project/traces/105445aa7843bc8bf206b12000100000",
        "logging.googleapis.com/spanId": "1",
        "logging.googleapis.com/traceSampled": true
}
```

* note: examples source/report locations are affected by the default stack filter so the logging module isn't always annotated as the log source. Note how logs written from within the handler's lifecycle have the extra annotations, trace id, etc. as the equivalent "now serving", and the "now serving" log entry can retain custom fields appended within the lifecycle of the request.

File: middleware.go [hunk 1]
```diff
@@ -1,19 +1,51 @@
...
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
 )
 
+// ProtoJSONMarshalOptions defines how we want to serialize request/response/status messages to logs
+// can be overridden
+var ProtoJSONMarshalOptions = protojson.MarshalOptions{
```
Comment (natefinch) [inline middleware.go]: global variables are evil

File: middleware.go [hunk 2]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+		Method: info.FullMethod,
+	})
+
+	resp, err = handler(ctx, req)
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
+
+	l.log(ctx, err, info.FullMethod)
+
+	return
```
Comment (natefinch) [inline middleware.go]: no.

```suggestion
	return resp, err
```

File: middleware.go [hunk 3]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+	startTime := time.Now()
+	ctx = ctxlogrus.ToContext(ctx, logrus.NewEntry(l.logger))
+
+	request := l.requestFromContext(ctx, &GRPCRequest{
+		Method: info.FullMethod,
+	})
+
+	resp, err = handler(ctx, req)
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
```
Comment (natefinch) [inline middleware.go]: why not milliseconds?  I hope we don't really need values over 1000 very often

File: middleware.go [hunk 4]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+	Deadline  string `json:"deadline,omitempty"`
+	Duration  string `json:"duration,omitempty"`
+}
+
+func (l loggingInterceptor) intercept(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (resp interface{}, err error) {
```
Comment (natefinch) [inline middleware.go]:
```suggestion
func (l loggingInterceptor) intercept(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
```

File: middleware.go [hunk 5]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+	startTime := time.Now()
+	ctx = ctxlogrus.ToContext(ctx, logrus.NewEntry(l.logger))
+
+	request := l.requestFromContext(ctx, &GRPCRequest{
+		Method: info.FullMethod,
+	})
+
+	resp, err = handler(ctx, req)
```
Comment (natefinch) [inline middleware.go]:
```suggestion
	resp, err := handler(ctx, req)
```

File: middleware.go [hunk 6]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+
+	return
+}
+
+func (l loggingInterceptor) interceptStream(
+	srv interface{},
+	ss grpc.ServerStream,
+	info *grpc.StreamServerInfo,
+	handler grpc.StreamHandler,
+) (err error) {
```
Comment (natefinch) [inline middleware.go]:
```suggestion
func (l loggingInterceptor) interceptStream(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, 	handler grpc.StreamHandler) (error) {
```

File: middleware.go [hunk 7]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+	request := l.requestFromContext(ctx, &GRPCRequest{
+		Method: info.FullMethod,
+	})
+
+	wrapped := grpc_middleware.WrapServerStream(ss)
+	wrapped.WrappedContext = ctx
+
+	err = handler(srv, wrapped)
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
```
Comment (natefinch) [inline middleware.go]: milliseconds is better I think

File: middleware.go [hunk 8]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+	ctx := ctxlogrus.ToContext(ss.Context(), logrus.NewEntry(l.logger))
+
+	request := l.requestFromContext(ctx, &GRPCRequest{
+		Method: info.FullMethod,
+	})
+
+	wrapped := grpc_middleware.WrapServerStream(ss)
+	wrapped.WrappedContext = ctx
+
+	err = handler(srv, wrapped)
```
Comment (natefinch) [inline middleware.go]:
```suggestion
	err := handler(srv, wrapped)
```

File: middleware.go [hunk 9]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
+
+	l.log(ctx, err, info.FullMethod)
+
+	return
```
Comment (natefinch) [inline middleware.go]:
```suggestion
	return err
```

**Review by natefinch — COMMENTED** [review]
not done, just letting these settle before going on

File: middleware.go [hunk 10]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
-				fields["trace"] = traceHeader
+			if o.filterHTTP(r) {
+				// log the result
+				ctxlogrus.Extract(ctx).Infof("served HTTP %v %v", r.Method, r.URL)
 			}
+		})
+	}
+}
+
+func UnaryLoggingInterceptor(logger *logrus.Logger, opts ...MiddlewareOption) grpc.UnaryServerInterceptor {
```
Comment (natefinch) [inline middleware.go:81]: comments on exported thingies please

File: middleware.go [hunk 11]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+	err = handler(srv, wrapped)
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
+
+	l.log(ctx, err, info.FullMethod)
+
+	return
+}
+
+func (l *loggingInterceptor) requestFromContext(ctx context.Context, request *GRPCRequest) *GRPCRequest {
```
Comment (natefinch) [inline middleware.go]: we don't seem to ever user this with an existing grpcrequest...  is there some other use, or could we move the info.FullRequest thingy inside here?

Also, comment, please

File: middleware.go [hunk 12]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+func (l *loggingInterceptor) log(ctx context.Context, err error, method string) {
+	if !l.filterRPC(ctx, method, err) {
+		return
+	}
+
+	if err != nil {
+		st := status.Convert(err)
+		jsonStatus, merr := ProtoJSONMarshalOptions.Marshal(st.Proto())
+		if merr != nil {
+			ctxlogrus.Extract(ctx).WithError(merr).Errorf("error marshalling error status into log")
```
Comment (natefinch) [inline middleware.go]: should we return here?

Also, let's only log something at error if we think it indicates an important bug that deserves immediate attention. This is just logging an error about logging

```suggestion
			ctxlogrus.Extract(ctx).WithError(merr).Warnf("error marshalling error status into log")
```

File: middleware.go [hunk 13]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+			return
+		}
+
+		if st.Code() == codes.Internal {
+			ctxlogrus.Extract(ctx).WithError(err).Errorf("internal error response on RPC %s", method)
+			return
+		}
 	}
+
+	ctxlogrus.Extract(ctx).Infof("served RPC %v", method)
```
Comment (natefinch) [inline middleware.go]: it's super hard for me to grok when a giant chunk of the code is in an if statement. can you put this in an if err == nil  before all this?  then you can outdent the rest of this and it's simpler to see what's going on.

File: middleware.go [hunk 14]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+				return
+			}
+
+			switch t := e.(type) {
+			case string:
+				err = errors.New(t)
+			case error:
+				err = t
+			default:
+				err = fmt.Errorf("unknown error: %w", t)
```
Comment (natefinch) [inline middleware.go]: I don't think you can use %w if it's not an error....  also it's explicitly not an error so.... 
```suggestion
				err = fmt.Errorf("unknown panic value: %v", t)
```

File: middleware.go [hunk 15]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+			// write error back to client
+			jsonStatus, merr := ProtoJSONMarshalOptions.Marshal(stErr.Proto())
+			if merr != nil {
+				entry.WithError(merr).Errorf("error marshalling error status into log")
+				fmt.Fprint(w, `{"error": "server_error"}`)
+				return
+			}
+
+			if _, err := w.Write(jsonStatus); err != nil {
+				entry.WithError(err).Fatal("error encoding json server_error to ResponseWriter")
```
Comment (natefinch) [inline middleware.go]: uh... Fatal?
> Fatal logs a message at level Fatal on the standard logger then the process will exit with status set to 1.
```suggestion
				entry.WithError(err).Info("error writing json server_error to ResponseWriter")
```

File: middleware.go [hunk 16]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+				fmt.Fprint(w, `{"error": "server_error"}`)
```
Comment (natefinch) [inline middleware.go]: if the write above fails, this is gonna fail, too ¯\\_(ツ)_/¯

**Review by natefinch — COMMENTED** [review]
still going....

File: middleware.go [hunk 17]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+		stErr := errWithStack(ctx, err)
+		err = stErr.Err()
+		resp = nil
+	}()
+
+	return handler(ctx, req)
+}
+
+// StreamRecoveryInterceptor is an interceptor that recovers panics from
+// Streaming services and turns them into nicer gRPC errors
```
Comment (natefinch) [inline middleware.go]:
```suggestion
// Streaming services and turns them into nicer gRPC errors.
```

File: middleware.go [hunk 18]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+			ctxlogrus.Extract(ctx).WithError(err).Errorf("internal error response on RPC %s", method)
+			return
+		}
 	}
+
+	ctxlogrus.Extract(ctx).Infof("served RPC %v", method)
+}
+
+// RecoveryMiddleware recovers from panics in the HTTP handler chain, logging
+// an error for Error Reporting
```
Comment (natefinch) [inline middleware.go]:
```suggestion
// an error for Error Reporting.
```

File: middleware.go [hunk 19]
```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
...
+
+		stErr := errWithStack(ss.Context(), err)
+		err = stErr.Err()
+	}()
+
+	return handler(srv, ss)
+}
+
+func errWithStack(ctx context.Context, err error) *status.Status {
+	stack := debug.Stack()
```
Comment (natefinch) [inline middleware.go:336]: This doesn't seem like it would give us useful information... it would give us the stack of grpc middleware calls and that's it. That seems boring, and debug.Stack() is kind of heavyweight to be calling for not-useful info.

File: middleware.go [hunk 20]
```diff
@@ -1,19 +1,51 @@
...
+	"net/url"
+	"runtime/debug"
 	"strconv"
+	"strings"
+	"time"
 
 	"github.com/felixge/httpsnoop"
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	"github.com/grpc-ecosystem/go-grpc-middleware/logging/logrus/ctxlogrus"
+	"github.com/lithammer/shortuuid"
```
Comment (natefinch) [inline middleware.go]: let's only have one uuid dependency, yeah?

File: middleware.go (hunk 3 above)
Comment (ClaudiaJ) [reply middleware.go]: Copied straight from the source material:
https://github.com/StevenACoffman/logrus-stackdriver-formatter/blob/5eec3c1a5678addc707c79ecf92fe0b8d0f47ccd/middleware.go#L30

Theoretically GCP's log viewer can take correctly formatted latency field. 9 sigfigs is probably in excess, but it's taken directly from Google's docs on LogEntry's HttpRequest latency field:
https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#HttpRequest.FIELDS.latency

Though the HttpRequestContext used by ErrorEvent to enrich Error Reporting dashboard has no need for latency, and I've not yet seen if the LogEntry can be enriched directly with the HttpRequest in its jsonPayload field, such that you can get this neat little widget on the log entry like we have in load balancer logs:
image
(note the actual value underlying is 5 sigfigs and it is converted to ms for the purpose of display automagically)

I'd like to have those, but once we have tracing to correlate all our logs to a single request context, it might not matter so much.

tl;dr: lmk if you feel strongly about this I can reduce it to 5 sigfigs like load balancer seems to use.

File: middleware.go (hunk 7 above)
Comment (ClaudiaJ) [reply middleware.go]: See above re: the duration format prescribed by the protocol I'm copying here. If it's ever parsed and formatted for display, it'll need to be as Google defines, duration in seconds with up to 9 fractional digits terminated b `s`.

File: middleware.go (hunk 11 above)
Comment (ClaudiaJ) [reply middleware.go]: This was mostly me refactoring a few things out so I didn't have to write-twice for unary and stream RPC's, it originally meant to add more stuff from the respective `info` param, e.g. `IsClientStream` and `IsServerStream`, but during refactoring I realized its superfluous and I got rid of it but never reduced this function to just take the FullMethod field.

I'll simplify it further and comment yeah, 👍

File: middleware.go (hunk 12 above)
Comment (ClaudiaJ) [reply middleware.go]: Yep good call, fixing that

File: middleware.go (hunk 13 above)
Comment (ClaudiaJ) [reply middleware.go]: I can add a comment in front of the error block.

I still want to log this even if error is not nil.
I'll rearrange it and add comments so it reads more clearer, but my purpose and intention here is to first add the gRPC Status to the log entry.
Only if the error that would be returned to the client is an internal error do I want it logging an Error.

Further I later added a custom error handler that I ended up not using in our services consuming this log handler, because it occurred to me we have the errors middleware that already converts everything to gRPC status without any of the additional flags and error context we add through the lifecycle of a request. This is where I had meant to do custom error handling, and we can still do that later replacing the errors middleware if we're so inclined.
But anyway the custom error handler is given the capability of logging for itself, returning truthy, and skipping the info-level logging (because it's been 'handled' already)

tl;dr: I'll try a different approach for parseability and lmk if that works or if you've got other ideas in the hat given what I'm trying to achieve here?

File: middleware.go (hunk 1 above)
Comment (ClaudiaJ) [reply middleware.go]: I invoke prior art:
https://github.com/grpc-ecosystem/go-grpc-middleware/blob/46f2eb369b917e60df91057c4d37847d17e5a9a4/logging/logrus/payload_interceptors.go#L16-L20

but ok sure fine I'll axe it.

File: middleware.go (hunk 16 above)
Comment (ClaudiaJ) [reply middleware.go]: my innate optimism showing through...

File: middleware.go (hunk 3 above)
Comment (ClaudiaJ) [reply middleware.go]: I tested real quick promoting the httpRequest to the parent Entry to see if it Just Works. It does:
image

But it overwrites the logged message with the requestURL, so I'm going to make a change that can promote this up only on these last "served {HTTP|RPC} [...]" entries.

Comment (ClaudiaJ) [reply middleware.go]: image

Made it work, even for gRPC (by pretending it's close enough to HTTP)

File: middleware.go (hunk 13 above)
Comment (ClaudiaJ) [reply middleware.go]: ok ignore all that rambling above, I improved it by moving that block into `handleError` that itself calls the custom error handler so now it's just an if handled then return

File: middleware.go (hunk 3 above)
Comment (natefinch) [reply middleware.go]: see... 0.000246141s is really hard for me to grok.  I don't like having to count zeroes to understand what a number means.  0.24ms is so much easier to understand, even if sometimes that means it's 3000ms.  but if that's how stackdriver needs it formatted, then so be it.  If we can do ms instead, I would prefer it. 

I would honestly drop it to 4 places after the decimal if that's an option. Do we really care if it's 0.1ms vs 0.14ms ?  Pretty sure we don't.

File: middleware_options.go [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+	FilterRPC  func(ctx context.Context, fullMethod string, err error) bool
+	FilterHTTP func(r *http.Request) bool
+
+	// ErrorHandler should return true if the error provided has already been logged
+	ErrorHandler func(ctx context.Context, err error, method string) (handled bool)
+)
+
+// TODO: default filters (NO healthchecks!)
+
+func DefaultFilterRPC(_ context.Context, fullMethod string, _ error) bool {
```
Comment (natefinch) [inline middleware_options.go:65]: comments for the default ones please

**Review by natefinch — APPROVED** [review]
mostly LGTM

File: middleware.go (hunk 19 above)
Comment (ClaudiaJ) [reply middleware.go:336]: It only occurs if a panic is surfaced back up from a request in the scope of the middleware, and GCP Error Reporting requires `debug.Stack()`, else we get no information at all to work with.

image

I can tell pretty quickly reading this what gRPC or HTTP method panicked, what line it panicked on, and if Cloud Debug were at all worth using you could click right through to source.

File: middleware.go (hunk 3 above)
Comment (ClaudiaJ) [reply middleware.go]: For consistency with how many sigfigs load balancer outputs, I'd want at least 5. In the grand scheme of things, I don't even care unless it's more than 3s. But I think you can also use this same field as-formatted to filter greater than some millisecond value same as I do in LB, e.g.:
```
trace:*
(httpRequest.latency>="3000ms" OR
jsonPayload.latency>="3000ms")
```

finds any entry with a traceID taking longer than 3s
